slides/aml-13-022818-resampling-imbalanced-data/index.html

```diff
@@ -0,0 +1,774 @@
+
+# Edited Nearest Neighbours
+
+- Originally as heuristic for reducing dataset for KNN
+- Remove all samples that are misclassified by KNN
+from training data (mode) or that have any point
+from other class as neighbor (all).
+- “Cleans up” outliers and boundaries.
+
+???
+
+---
+
+.center[
+![:scale 90%](images/edited_nearest_neighbour.png)
+]
+
+???
+
+---
+
+# Edited Nearest Neighbours
+
+.smaller[
+```python
+from imblearn.under_sampling import EditedNearestNeighbours
+enn = EditedNearestNeighbours(n_neighbors=5)
+X_train_enn, y_train_enn = enn.fit_sample(X_train, y_train)
+
+enn_mode = EditedNearestNeighbours(kind_sel="mode", n_neighbors=5)
+X_train_enn_mode, y_train_enn_mode = enn_mode.fit_sample(X_train, y_train)
+```
+.center[
+![:scale 100%](images/edited_nearest_neighbour_2.png)
+]
+]
+
+???
+
+---
+
+.smaller[
+```python
+enn_pipe = make_imb_pipeline(EditedNearestNeighbours(n_neighbors=5),
+                             LogisticRegression())
+scores = cross_val_score(enn_pipe, X_train, y_train, cv=10, scoring='roc_auc')
+print(np.mean(scores))
+```
+0.920
+```python
+enn_pipe_rf = make_imb_pipeline(EditedNearestNeighbours(n_neighbors
+                      = 5),RandomForestClassifier(n_estimators=100))
+scores = cross_val_score(enn_pipe_rf, X_train, y_train, cv=10, scoring='roc_auc')
+print(np.mean(scores))
+```
+0.944
+]
+???
+
+---
+
+# Condensed Nearest Neigbours
+
+- Iteratively adds points to the data that are
+misclassified by KNN
+- Focuses on the boundaries
+- Usually removes many
+
+```python
+cnn = CondensedNearestNeighbour()
+X_train_cnn, y_train_cnn = cnn.fit_sample(X_train, y_train)
+print(X_train_cnn.shape)
+print(np.bincount(y_train_cnn))
+```
+(556, 6) </br>
+[361 195]
+
+???
+
+---
+
+.center[
+![:scale 100%](images/edited_condensed_nn.png)
+]
+
+???
+
+---
+
+.smaller[
+```python
+cnn_pipe = make_imb_pipeline(CondensedNearestNeighbour(), LogisticRegression())
+scores = cross_val_score(cnn_pipe, X_train, y_train, cv=10, scoring='roc_auc')
+print(np.mean(scores))
+```
+0.919
+```python
+cnn_pipe = make_imb_pipeline(CondensedNearestNeighbour(),
+                          RandomForestClassifier(n_estimators=100))
+scores = cross_val_score(cnn_pipe, X_train, y_train, cv=10, scoring='roc_auc')
+print(np.mean(scores))
+```
+0.948
+]
+???
+
+---
+class: center, middle
+
+## Synthetic Sample Generation
+
+???
+
+---
+
+# Synthetic Minority Oversampling Technique
+# (SMOTE)
+
+- Adds synthetic interpolated data to smaller class
+- For each sample in minority class:</br>
+  – Pick random neighbor from k neighbors.</br>
+  – Pick point on line connecting the two uniformly</br>
+  – Repeat.</br>
+- Leads to very large datasets (oversampling)
+- Can be combined with undersampling strategies
+
+
+???
+
+---
+
+.center[
+![:scale 100%](images/smote.png)
+]
+
+???
+
+---
+
+class: center, middle
+
+.center[
+![:scale 100%](images/smote_2.png)
+]
+
+???
+
+
+---
+
+class: center, middle
+
+.center[
+![:scale 100%](images/smote_3.png)
```

Comment:
thanks for recreating the image. We don't really need the one without labels, though (smote_2) and it would be nice to get labels on smote.png as well.

Reply:
Cool. Will do.